apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableReadRowsRequestManager.cs

```diff
@@ -0,0 +1,187 @@
+
+namespace Google.Cloud.Bigtable.V2
+{
+    /// <summary>
+    /// Keeps track of Rows returned from a readRows RPC for information relevant to resuming the RPC
+    /// after temporary problems.
+    /// </summary>
+    internal class BigtableReadRowsRequestManager
+    {
+        /// <summary>
+        /// Compares two <see cref="ByteString"/>s
+        /// </summary>
+        /// <returns>integer that indicates their relative position in the lexicographical order.</returns>
+        private int ByteStringCompare(ByteString key1, ByteString key2)
+        {
+            int size1 = key1.Length;
+            int size2 = key2.Length;
+            int size = Math.Min(size1, size2);
+
+            for (int i = 0; i < size; i++)
+            {
+                int comparison = key1[i].CompareTo(key2[i]);
+                if (comparison != 0)
+                {
+                    return comparison;
+                }
+            }
+            return size1.CompareTo(size2);
+        }
+
+        // Memeber variables form the constructor.
+        private readonly ReadRowsRequest _originalRequest;
+
+        // The number of rows read so far.
+        private long _rowCount;
+
+        /// <summary>
+        /// Constructor for ResumingStreamingResultScanner.
+        /// </summary>
+        /// <param name="originalRequest">
+        /// Original ReadRowsRequest containg all of the parameters of the API call.
+        /// </param>
+        public BigtableReadRowsRequestManager(ReadRowsRequest originalRequest)
+        {
+            _originalRequest = originalRequest;
+        }
+
+        /// <summary>
+        /// Update last found Rowkey
+        /// </summary>
+        public ByteString LastFoundKey { get; set; }
+
+        /// <summary>
+        /// Update amount of rows read so far
+        /// </summary>
+        public long RowCount
+        {
+            get => _rowCount;
+            set => _rowCount += value;
+        }
+
+        /// <summary>
+        /// Builds and returns updated subrequest
+        /// </summary>
+        /// <returns></returns>
+        public ReadRowsRequest BuildUpdatedRequest()
+        {
+            ReadRowsRequest newReadRowsRequest = new ReadRowsRequest()
+
+            {
+                TableNameAsTableName = _originalRequest.TableNameAsTableName,
+            };
+
+            #region Set up RowFilter
+
+            if (_originalRequest.Filter != null)
+            {
+                newReadRowsRequest.Filter = new RowFilter();
+                newReadRowsRequest.Filter = _originalRequest.Filter;
+            }
+
+            #endregion
+
+            #region Set up RowsLimit
+
+            // If the row limit is set, update it.
+            long numRowsLimit = _originalRequest.RowsLimit;
+
+            // Update <see cref="numRowsLimit"/> by removing the number of rows already read.
+            if ((numRowsLimit -= RowCount) > 0)
+            // The remaining number of rows must be greater than 0.
```

Comment:
Nit: I would move the comment above the if condition so it doesn't break up the block.

However, bigger issue: I'm assuming this comment means we would never get into a situation where the number of remaining rows is zero, right? If so, I would be a bit more _assertive_ here (apologies for the attempt at wordplay humor):
```csharp
if (_originalRequest.RowsLimit != 0)
{
    long rowsRemaining = _originalRequest.RowsLimit - RowCount;
    Debug.Assert(rowsRemaining > 0, "The remaining number of rows must be greater than 0.");
    newReadRowsRequest.RowsLimit = Math.Max(1, rowsRemaining);
}
```
That way if the assumption is ever wrong for some unforeseen reason, we'll realize it when debugging and in release builds we won't start firing off followup ReadRows requests that are unlimited.